<a href="https://colab.research.google.com/github/u-adrian/PSDA/blob/main/Aufgabe_5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


# Aufgabe 5: Interpretation von Modellen

a) Lesen Sie den Datensatz Hdma.csv ein und machen Sie sich mit den Daten vertraut.

In [2]:
import pandas as pd
import os


In [3]:
DATA_PATH = "/content/drive/MyDrive/PDSA/ÜB1/dataset"

In [4]:
df = pd.read_csv(os.path.join(DATA_PATH,"Hdma.csv"),sep=';')

In [5]:
df.head()

dir    hir       lvr  ccs  mcs pbcr dmi self single  uria  condo black  \
0  0.221  0.221  0.800000  5.0  2.0   no  no   no     no   3.9      0    no   
1  0.265  0.265  0.921875  2.0  2.0   no  no   no    yes   3.2      0    no   
2  0.372  0.248  0.920398  1.0  2.0   no  no   no     no   3.2      0    no   
3  0.320  0.250  0.860465  1.0  2.0   no  no   no     no   4.3      0    no   
4  0.360  0.350  0.600000  1.0  1.0   no  no   no     no   3.2      0    no   

  deny  
0   no  
1   no  
2   no  
3   no  
4   no

###Variable Description and More Evident Renaming


| Original Variable     | Description | New Variable |
| ----------- | ----------- | ----------- |
| dir     | debt payments to total income ratio      | debt_payments_to_total_income_ratio      |
| hir   | housing expenses to income ratio        | housing_expenses_to_income_ratio      |
| lvr      | ratio of size of loan to assessed value of property      | size_of_loan_to_assessed_value_of_property_ratio      |
| ccs   | consumer credit score from 1 to 6 (a low value being a good score)        | consumer_credit_score      |
| mcs      | mortgage credit score from 1 to 4 (a low value being a good score)      | mortgage_credit_score_from      |
| pbcr   | public bad credit record?        | public_bad_credit_record      |
| dmi   | denied mortgage insurance ?       | denied_mortgage_insurance      |
| self      | self employed?       | self_employed      |
| single   | is the applicant single?      | single      |
| uria      | 1989 Massachusetts unemployment rate in the applicant’s industry       | unemployment_rate_in_the_applicant’s_industry      |
| condominium   | is unit a condominium?        | condominium      |
| black      | is the applicant black?       | black      |
| deny   | mortgage application denied?        | deny      |

In [6]:
new_column_names = ["_".join("debt payments to total income ratio".split()),
"_".join("housing expenses to income ratio".split()),
 "_".join("size of loan to assessed value of property ratio".split()),
 "_".join("consumer credit score".split()),
 "_".join("mortgage credit score".split()),
 "_".join("public bad credit record".split()),
 "_".join("denied mortgage insurance".split()),
 "_".join("self employed".split()),
 "_".join("single".split()),
 "_".join("unemployment rate in the applicant’s industry".split()),
 "_".join("condominium".split()),
 "_".join("black".split()),
 "_".join("deny".split())]


df.columns = new_column_names

In [7]:
df.head()

debt_payments_to_total_income_ratio  housing_expenses_to_income_ratio  \
0                                0.221                             0.221   
1                                0.265                             0.265   
2                                0.372                             0.248   
3                                0.320                             0.250   
4                                0.360                             0.350   

   size_of_loan_to_assessed_value_of_property_ratio  consumer_credit_score  \
0                                          0.800000                    5.0   
1                                          0.921875                    2.0   
2                                          0.920398                    1.0   
3                                          0.860465                    1.0   
4                                          0.600000                    1.0   

   mortgage_credit_score public_bad_credit_record denied_mortgage_insurance  \
0                    2.0                       no                        no   
1                    2.0                       no                        no   
2                    2.0                       no                        no   
3                    2.0                       no                        no   
4                    1.0                       no                        no   

  self_employed single  unemployment_rate_in_the_applicant’s_industry  \
0            no     no                                            3.9   
1            no    yes                                            3.2   
2            no     no                                            3.2   
3            no     no                                            4.3   
4            no     no                                            3.2   

   condominium black deny  
0            0    no   no  
1            0    no   no  
2            0    no   no  
3            0    no   no  
4            0    no   no

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2381 entries, 0 to 2380
Data columns (total 13 columns):
 #   Column                                            Non-Null Count  Dtype  
---  ------                                            --------------  -----  
 0   debt_payments_to_total_income_ratio               2381 non-null   float64
 1   housing_expenses_to_income_ratio                  2381 non-null   float64
 2   size_of_loan_to_assessed_value_of_property_ratio  2381 non-null   float64
 3   consumer_credit_score                             2381 non-null   float64
 4   mortgage_credit_score                             2381 non-null   float64
 5   public_bad_credit_record                          2380 non-null   object 
 6   denied_mortgage_insurance                         2381 non-null   object 
 7   self_employed                                     2380 non-null   object 
 8   single                                            2381 non-null   object 
 9   unemployment_rate_i

In [9]:
df.describe()

debt_payments_to_total_income_ratio  housing_expenses_to_income_ratio  \
count                          2381.000000                       2381.000000   
mean                              0.330814                          0.255346   
std                               0.107235                          0.096635   
min                               0.000000                          0.000000   
25%                               0.280000                          0.214000   
50%                               0.330000                          0.260000   
75%                               0.370000                          0.298800   
max                               3.000000                          3.000000   

       size_of_loan_to_assessed_value_of_property_ratio  \
count                                       2381.000000   
mean                                           0.737760   
std                                            0.178715   
min                                            0.020000   
25%                                            0.652850   
50%                                            0.779412   
75%                                            0.868421   
max                                            1.950000   

       consumer_credit_score  mortgage_credit_score  \
count            2381.000000            2381.000000   
mean                2.116387               1.721008   
std                 1.666370               0.537169   
min                 1.000000               1.000000   
25%                 1.000000               1.000000   
50%                 1.000000               2.000000   
75%                 2.000000               2.000000   
max                 6.000000               4.000000   

       unemployment_rate_in_the_applicant’s_industry  condominium  
count                                    2381.000000  2381.000000  
mean                                        3.774496     0.288114  
std                                         2.026636     0.452980  
min                                         1.800000     0.000000  
25%                                         3.100000     0.000000  
50%                                         3.200000     0.000000  
75%                                         3.900000     1.000000  
max                                        10.600000     1.000000

b) Volkswirtschaftler möchten herausfinden, was die möglichen Einflussfaktoren auf die
Ablehnung eines Kreditantrags sind, und ob es eine rassistische Diskriminierung gibt. Welcher Typ von Aufgabe des maschinellen Lernens ist für diese Aufgabe geeignet? Welche
Methoden kann man benutzen?

**Antwort:** 
Grundsätzlich kann die Aufgabe der Feature Selection im Maschninellen Lernen Auskunft darüber geben wie groß der Einflussfaktor einzelner Features auf die Target Variable ist. Zum Beispiel eignen sich die Klassen im [sklearn.feature_selection](https://scikit-learn.org/stable/modules/feature_selection.html#feature-selection) Modul um disen Zusammenhang zu untersuchen. Eine Methode um hiermit den Zusammenhang zwischen Featuren und der Target Variable zu untersuchen ist die Nutzung von statistischen Tests. Die Art des anzuwendenden statistischen Tests richtet sich nach dem Datentyp (categorical, ordinal und numeric) des Features (siehe Aufgabe 4). 

Eine weitere Methode ist pandas interne DataFrame.corr() Funktion mit der man die paarweise Korrelation zwischen Spalten bestimmen kann. Diese wird typischerweise ebenfalls im Feature Selection Prozess angewendet. Die gewünschte Korrelationsmethode (Pearson, Spearman, Kendalls Tau) kann hierbei gewählt werden.

Eine weitere Methode findet sich in der Betrachtung von Decision Tree Algorithmen. Hierbei kann das Auswahlkriterium zur Aufteilung der Daten (z.B. der Information Gain) eines Features genutzt werden um Aussagen über den Einfluss dieses Features auf die Target Variable zu machen. Decision Trees sind nämlich genau nach dem Prinzip aufgebaut, dass die Features mit dem größten Einfluss näher an der Wurzel lokalisiert sind. Beispielsweise kann man in sklearn über das feature_importances_ Attribut eines Entscheidungsbaum auf diese Größe zugreifen. 



Mit den oben vorgestellten Methoden kann der Zusammenhang zwischen dem Feature **black** und dem Target **deny** ermittelt werden. Daraus kann man ableiten ob eine rassistische Diskrimierung vorliegt.


In die Kontextsensitive Systeme Vorlesun reingucken!

Furthermore, impurity-based feature importance for trees are strongly biased and favor high cardinality features (typically numerical features) over low cardinality features such as binary features or categorical variables with a small number of possible categories.

Ggf. noch Permutation Feature Importance aus Sklearn mitaufnehmen: https://scikit-learn.org/stable/modules/permutation_importance.html#permutation-importance

c) Ist Datenvorverarbeitung nötig? Warum? Falls nötig, verarbeiten Sie die Daten vor.

**Antwort:** <br> Für einige der statistischen Tests, für die DataFrame.corr() Funktion und viele der ALG in Teilaufgabe f) ist es notwendig die nominalen Daten mit den Ausprägungen "yes" und "no" in numerische Daten ("0" bzw. "1") umzuwandeln. Dies trifft auf folgende Features zu: **public_bad_credit_record, denied_mortgage_insurance, self_employed, single, unemployment_rate_in_the_applicant’s_industry, black**. Und das Target: **deny**. <br><br> Darüberhinaus sollten die Daten skaliert werden, da Unterschiede in dem Ausprägungsbereich der Daten Einfluss auf die Performanz der ML-Algorithmen haben können. Dies ist für diese Aufgabe allerdings nicht unmittelbar relevant, da wir in den folgenden Teilaufgaben nur an den relativen Unterschieden zwischen den Featuren Einflüssen interessiert sind und nicht an der Gesamtperformance der ALG. Daher werden wir uns in diesem Abschnitt nicht tiefer mit der Auswahl der Skalierungsmethode beschäftigen, sondern die Daten lediglich standardisieren.

Muss man für die Anwendung von Decision Trees, die Daten in Intervalle einteilen?

Feaute Importance is misleading for features that have many unique values! https://scikit-learn.org/stable/auto_examples/ensemble/plot_forest_importances.html Therefore we need Data Preparation

In [10]:
#Transformieren von nominalen Daten in numerische Daten

yes_no_to_int = {"yes": 1, "no": 0}

for column in df.columns:
 if df[column].dtype == "object":
    df[column] = df[column].map(yes_no_to_int)
   


In [11]:
#Daten standardisieren
# ist eigentlich nicht sinnvoll..


from sklearn.preprocessing import StandardScaler
standard_scaler = StandardScaler()
data_standardized = standard_scaler.fit_transform(df)
df_standardized = pd.DataFrame(data_standardized)


d) Separieren Sie die Daten in 80% : 20% = Training : Test Datensatz. Führen Sie eine logistische 
Regression durch. Beschreiben Sie Koeffizienten und deren Signifikanzniveau sowie die Area 
Under Curve (AUC) der Receiver-Operator-Curve (ROC). Interpretieren Sie die Ergebnisse.

In [12]:
#Daten in Feature und Target separieren

X = df.iloc[:, :-1]
y = df.iloc[:, -1]



In [13]:
# Aufteilen der Daten in einen Trainings- und einen Testdatensatz

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0 )


In [57]:
# Benutzung von Phyton's statsmodels Modul um eine logistische Regression durchzuführen und statistische Metriken auszugeben
import statsmodels.api as sm
log_regressor = sm.Logit(y_train, sm.add_constant(X_train)).fit()


Optimization terminated successfully.
         Current function value: 0.275621
         Iterations 7


/usr/local/lib/python3.7/dist-packages/statsmodels/tsa/tsatools.py:117: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


In [58]:
#Darstellung der Parameter des logistischen Regressionsmodell und zugehörigen Signifikanzwerten
test = log_regressor.summary().tables[1].as_html()
dff = pd.read_html(test, header=0, index_col=0)
dff[0].sort_values(by="coef", ascending=False)


coef  std err       z  \
denied_mortgage_insurance                         4.2408    0.561   7.554   
debt_payments_to_total_income_ratio               4.1928    1.134   3.696   
size_of_loan_to_assessed_value_of_property_ratio  1.5981    0.534   2.992   
public_bad_credit_record                          1.2776    0.229   5.581   
black                                             0.7478    0.198   3.781   
self_employed                                     0.6548    0.236   2.775   
housing_expenses_to_income_ratio                  0.4760    1.376   0.346   
single                                            0.4170    0.173   2.410   
consumer_credit_score                             0.2962    0.044   6.755   
mortgage_credit_score                             0.1589    0.162   0.982   
unemployment_rate_in_the_applicant’s_industry     0.0626    0.038   1.652   
condominium                                       0.0435    0.188   0.232   
const                                            -6.8219    0.608 -11.214   

                                                  P>|z|  [0.025  0.975]  
denied_mortgage_insurance                         0.000   3.141   5.341  
debt_payments_to_total_income_ratio               0.000   1.969   6.416  
size_of_loan_to_assessed_value_of_property_ratio  0.003   0.551   2.645  
public_bad_credit_record                          0.000   0.829   1.726  
black                                             0.000   0.360   1.136  
self_employed                                     0.006   0.192   1.117  
housing_expenses_to_income_ratio                  0.729  -2.221   3.173  
single                                            0.016   0.078   0.756  
consumer_credit_score                             0.000   0.210   0.382  
mortgage_credit_score                             0.326  -0.158   0.476  
unemployment_rate_in_the_applicant’s_industry     0.099  -0.012   0.137  
condominium                                       0.817  -0.324   0.411  
const                                             0.000  -8.014  -5.630



###Beschreiben der Koeffizienten und deren Signifikanzniveaus
...





### Koeffizienten
Die Spalte **coef** der log_regressor.summary() Ausgabe stellt die Koeffizienten des logistischen Regressionsmodells dar. Um die Koeffizienten zu verstehen betrachten wir zunächst ein vereinfachtes logistische Regressionsmodell, welches nur die Attribute **single** und **black** betrachtet: <br><br> $log(p/1-p)=0.4170*single+0.7478*black$ <br><br> $p = P(Y=1)$ entspricht der Wahrscheinlichkeit, dass der Kreditantrag abgelehnt wird. Daraus folgt, dass $1-p = P(Y=0)$ die Wahrscheinlicheit darstellt, dass der Kreditantrag angenommen wird. Das Verhältnis der Wahrscheinlichkeit für die Ablehnung zur Wahrschieinlichkeit für die Annahme $P(Y=1)/P(Y=0)$ wird auch **Chancenverhältniss** genannt. <br><br>
Angenommen das Attribut **single** hat den Wert 1 für eine Beobachtung, dann erhöht sich $log(p/1-p)$ um den Koeffizienten von **single**, nämlich 0.4170 in unserem Fall. Das heißt, dass sich das **Chancenverhältniss** um $exp(0.4170)=1.5174$ erhöht. Und das bedeutet, dass die Chance (Wahrscheinlichkeit) für eine Kreditablehnung um 51.74% steigt im Verhältniss zur Chance der Kreditannahme. <br><br> Generell lässt sich sagen: wenn das Vorzeichen des Koeffizienten eines Attributes positiv ist, dann wird die Wahrscheinlichkeit für eine Kreditablehnung größer. Ist das Vorzeichen negativ, dann wird die Wahrscheinlichkeit für eine Kreditannahme größer.

### Signifikanzniveau
Die Einschätzung der Signifikanz der ermittelten Koeffizienten der Attribute des logistischen Regressionsmodels kann mit Hilfe des p-Wertes erfolgen, welcher einem Hypothesentest zur Grund liegt.
Das Modul statsmodels ermöglicht es uns für jeden ermittelten Koeffizienten einen Z-Hypothesentest durchzuführen, mit dem getestet wird mit welcher Sicherheit der Koeffizient ungleich 0 ist. Genauer gesagt ist die H0: coef==0, die H1: coef!=0.
Die Spalte  **P>|z|** stellt den Signifikanzwert(p-Wert) dar, welcher beschreibt in wie weit die Nullhypothese H0 glaubwürdig ist. Ein niedriger p-Wert bedeutet, dass die Nullhypothese H0 zu gunsten der Alternativhypothese H1 verworfen werden kann. Welcher p-Wert hierfür genau gewählt wird ist Kontext abhänig. Generell folgt, dass je niedriger der p-Wert ist, desto sicherer ist die Alternativhypohtese und demnach die Güte des ermittelten Koeffizienten. In unserem Kontext kann zu dem folgende Interpreation gemacht werden: betrachtet man den p-Wert von dem Attribut **single** mit **P>|z|=0.016**, dann kan daraus schließen, dass es eine 1.6% Chance gibt, dass das Attribut **single** keinen Einfluss auf unser target **deny** hat [(Quelle für diese Interpretation)](https://medium.com/swlh/interpreting-linear-regression-through-statsmodels-summary-4796d359035a). 



###Beschreiben der Koeffizienten und deren Signifikanzniveaus
...

### Beschreiben von Area Under Curve (AUC) der Receiver-Operator-Curve (ROC)


### Ergebnisse Interpretieren
gerade mit Bezug zu Teilaufgabe b) rassistischer Zusammenhang?








